lib/std/meta.zig

```diff
@@ -965,18 +965,34 @@ test "intToEnum with error return" {
         A,
         B,
     };
+    const E3 = enum(u8) {
+        A,
+        _
+    };
 
     var zero: u8 = 0;
     var one: u16 = 1;
     try testing.expect(intToEnum(E1, zero) catch unreachable == E1.A);
     try testing.expect(intToEnum(E2, one) catch unreachable == E2.B);
+    try testing.expect(intToEnum(E3, zero) catch unreachable == E3.A);
+    try testing.expect(intToEnum(E3, 255) catch unreachable == @intToEnum(E3, 255));
     try testing.expectError(error.InvalidEnumTag, intToEnum(E1, one));
+    try testing.expectError(error.InvalidEnumTag, intToEnum(E3, 256));
 }
 
 pub const IntToEnumError = error{InvalidEnumTag};
 
 pub fn intToEnum(comptime EnumTag: type, tag_int: anytype) IntToEnumError!EnumTag {
-    inline for (@typeInfo(EnumTag).Enum.fields) |f| {
+    const enum_info = @typeInfo(EnumTag).Enum;
+
+    if (!enum_info.is_exhaustive) {
+        if (tag_int > std.math.maxInt(enum_info.tag_type)) {
```

Comment:
This should also handle underflow. Maybe use `std.math.cast`.

Reply:
Hmm I think the logic in the PR is incorrect since enums are not necessarily contiguous. Here's another test case to add to this PR:

```zig
const E = enum {
    a = 10,
    b = 100,
    c = 1000,
    _,
};
test "nonexhaustive with holes" {
    try expectError(error.InvalidEnumTag, std.meta.intToEnum(E, 9));
    try expectError(error.InvalidEnumTag, std.meta.intToEnum(E, 11));
    try expectError(error.InvalidEnumTag, std.meta.intToEnum(E, 1001));
    try expectEqual(E.a, std.meta.intToEnum(E, 10));
    try expectEqual(E.b, std.meta.intToEnum(E, 100));
    try expectEqual(E.c, std.meta.intToEnum(E, 1000));
}
```

Reply:
It shouldn't matter whether or not the named enum values are contiguous. If the enum is non-exhaustive, any value in the range of its tag integer is valid (i.e. `@enumToInt(enum(u8) { a = 1, b = 3 _ }, 2)` is well-defined).

Reply:
The use case you are describing is handled by directly calling `@intToEnum`, in which case there is no purpose of calling `std.meta.intToEnum`.

Reply:
Whoops, fixed for signed ints!

@andrewrk generic code may want to call `std.meta.intToEnum` without checking whether the enum is exhaustive or not. (e.g. [std.json does it here.](https://github.com/ziglang/zig/blob/fd6200eda6d4fe19c34a59430a88a9ce38d6d7a4/lib/std/json.zig#L1504))

The way I see it there are two options:
1. `std.meta.intToEnum` behaves equivalently to `@intToEnum`, but on bad input returns an error code instead of tripping a safety check. That's what this PR implements.
2. `std.meta.intToEnum` only returns named variants, its behavior is distinct from that of `@intToEnum`. It's up to the caller to check if the enum is exhaustive. If I understand correctly that's what you have in mind.

Either are self consistent, but in my opinion 1) is easier to work with, and 2) is surprising behavior for the current function name.

Reply:
Sure, not directly, but the original use case being addressed by this change is:
> This was previously preventing std.json from deserializing non-exhaustive enums because std.meta.intToEnum would consider non-named enum fields invalid.

Which, I at least imagine, would include casting from unnamed values in between named values. Though, if that lenient behavior isn't desirable, then this change shouldn't be made in any capacity, and instead the json code should be made to do this check instead.